Adds Zero API guide (#1438)
pomerium/documentation: content/docs/zero/api-guide.md

### Get organization ID

There are two types of organizations: **personal** and **professional**. When you create an account, you get a personal organization by default.

To get your personal organization ID, send a `GET` request to the `/organizations` endpoint:

```curl
curl --location 'https://console.pomerium.app/api/v0/organizations' \
--header 'Authorization: Bearer <ID-TOKEN>' \
--data ''
```

If your request was successful, you'll get a response with your organization ID and other related metadata:

```json
[
  {
    "createdAt": "2024-01-17T20:07:47.794672Z",
    "id": "bwjkRZwxbNXBQHHcJHphGSNBbxt",
    "joinedAt": "2024-05-29T20:27:27.336939Z",
    "name": "personal",
    "organizationType": "personal",
    "ownerUserId": "bMWxgvnRQfcRHjQvfMehbJhYFzB",
    "role": "admin",
    "updatedAt": "2024-01-17T20:07:47.794672Z"
  }
]
```

Copy your organization ID.

### Get namespace ID

In Pomerium Zero, each cluster is assigned its own namespace. To make changes to your cluster's configuration, you need your cluster's namespace ID.

To get your cluster's namespace ID, send a `GET` request to the `/namespaces` endpoint, replacing `{organizationId}` with your own:

```curl

> curl --location 'https://console.pomerium.app/api/v0/organizations/{organizationId}/namespaces' \

Reviewer: ```suggestion
curl --location 'https://console.pomerium.app/api/v0/organizations/{organizationId}/clusters' \
```

Maybe we should list clusters as this is the concept available in the API, while namespace is currently not part of the UI concepts; the cluster entity would have `namespaceId` that user could use in the subsequent calls to create routes and policies

Author: I'm incorporating this feedback and removing the previous payload example. I'll also remove the parent/child cluster relationship explanation.

--header 'Authorization: Bearer <ID-TOKEN>'
```

> You'll get a response similar to the example below. In the JSON payload, the first object contains data about the organization, which functions as a root namespace with admin-level permissions.

Reviewer: ```suggestion
You'll get a response similar to the example below. In the JSON payload, the first object contains data about the cluster, which has an ID of a root namespace other entities such as routes and policies are assigned to.
```

  },
  {
    "createdAt": "2024-01-17T20:07:49.173932Z",
    "id": "bjQWGxKgPJRSZWSBWpgRWMpJbsv",
    "name": "trusted-dog-1049",
    "parentId": "bwjkRZwxbNXBQHHcJHphGSNBbxt",
    "role": "admin",
    "type": "cluster",
    "updatedAt": "2024-05-29T15:07:07.938956Z"
  }
]
```

Go ahead and copy the `"cluster"` namespace ID.

### Build a policy

Now, you'll build a policy that only grants access if the user's email address matches exactly the email address specified in the policy.

To create a policy, send a `POST` request to the `/policies` endpoint:

```curl
curl --location 'https://console.pomerium.app/api/v0/organizations/{organizationId}/policies' \
--header 'Content-Type: application/json' \
--header 'Authorization: <ID-TOKEN>' \
--data-raw '{
  "namespaceId": "bwjkRZwxbNXBQHHcJHphGSNBbxt",
  "name": "Allow only matching email",
  "enforced": false,
  "ppl": {
    "allow": {
      "and": [
        {"email":
            {
            "is": "user@example.com"
            }
        }
      ]
    }
  },
  "description": "Only allow users access with a matching email address.",
  "explanation": "You don'\''t have the correct email address to access this service.",

>   "remediation": "Email address must match user@example.com"

Reviewer: ```suggestion
  "remediation": "Use account credentials that match policy requirements."
```

let's not disclose the content of authorization policy

}'
```

:::info

In the request example above, setting the `"enforced"` field to `false` means that the policy won't be automatically assigned to routes in your cluster. Setting this field to `true` has the opposite effect: it will assign this policy to your existing routes and any new routes added to your cluster.

:::

If your request was successful, you should receive a response similar to the one below. Notice that the `"routes"` field is empty because this policy hasn't been assigned yet:

```json
{
  "createdAt": "2024-05-30T17:04:33.460441Z",
  "description": "Only allow users with a matching email address.",
  "enforced": false,
  "enforcedRoutes": [],
  "explanation": "You don't have the correct email address to access this service.",
  "id": "bgrXNgrJFJmMZvPsVsbZHGWxVWP",
  "name": "Allow matching email",
  "namespaceId": "bwjkRZwxbNXBQHHcJHphGSNBbxt",
  "ppl": {
    "allow": {
      "and": [
        {
          "email": {
            "is": "user@example.com"
          }
        }
      ]
    }
  },

>   "remediation": "Email address must match user@example.com",

Reviewer: ```suggestion
  "remediation": "Use account credentials that match policy requirements.",
```

  "routes": [],
  "updatedAt": "2024-05-30T17:04:33.460441Z"
}
```

## 3. Create a route

Next, you'll create a route and attach it to the policy you just created. In this example, we're using [HTTPBin](https://httpbin.org/) as an example service, but you can use any service you want.

To create a route, you need to send a `POST` request to the `/routes` endpoint with your organization ID, namespace ID, and the policy ID of the policy you just created. Make sure you replace `{CLUSTER_STARTER_SUBDOMAIN}` in the `"from"` field with your own:

```curl
curl --location 'https://console.pomerium.app/api/v0/organizations/{organizationId}/routes' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer {ID-TOKEN}' \
--data '{
    "namespaceId": "bwjkRZwxbNXBQHHcJHphGSNBbxt",
    "name": "HTTP Bin",
    "from": "https://httpbin.{CLUSTER_STARTER_SUBDOMAIN}.pomerium.app",
    "to": ["http://httpbin:80"],

>     "allowWebsockets": false,
>     "allowSpdy": false,
>     "tlsSkipVerify": false,
>     "tlsUpstreamAllowRenegotiation": false,
>     "preserveHostHeader": false,
>     "enableGoogleCloudServerlessAuthentication": false,
>     "showErrorDetails": false,

Reviewer: ```suggestion
```
